pkg/scheduler/schedule_one.go

```diff
@@ -88,19 +88,49 @@ func (sched *Scheduler) scheduleOne(ctx context.Context) {
 	start := time.Now()
 	state := framework.NewCycleState()
 	state.SetRecordPluginMetrics(rand.Intn(100) < pluginMetricsSamplePercent)
+
 	// Initialize an empty podsToActivate struct, which will be filled up by plugins or stay empty.
 	podsToActivate := framework.NewPodsToActivate()
 	state.Write(framework.PodsToActivateKey, podsToActivate)
 
 	schedulingCycleCtx, cancel := context.WithCancel(ctx)
 	defer cancel()
-	scheduleResult, err := sched.SchedulePod(schedulingCycleCtx, fwk, state, pod)
+	scheduleResult, assumedPodInfo := sched.schedulingCycle(schedulingCycleCtx, state, fwk, podInfo, podsToActivate, start)
+	if scheduleResult.FeasibleNodes == 0 {
+		return
+	}
+
+	// bind the pod to its host asynchronously (we can do this b/c of the assumption step above).
+	go func() {
+		bindingCycleCtx, cancel := context.WithCancel(ctx)
+		defer cancel()
+
+		metrics.SchedulerGoroutines.WithLabelValues(metrics.Binding).Inc()
+		defer metrics.SchedulerGoroutines.WithLabelValues(metrics.Binding).Dec()
+
+		sched.bindingCycle(bindingCycleCtx, state, fwk, scheduleResult, assumedPodInfo, podsToActivate, start)
+	}()
+}
+
+// schedulingCycle tries to schedule a single Pod.
+func (sched *Scheduler) schedulingCycle(ctx context.Context, state *framework.CycleState, fwk framework.Framework, podInfo *framework.QueuedPodInfo, podsToActivate *framework.PodsToActivate, start time.Time) (_ ScheduleResult, retPodInfo *framework.QueuedPodInfo) {
+	var retErr error
```

Comment:
Just my poor point, in order to reduce the return values, we add three variables in this function and we have to assign them each time when return, then we defer to handle the error. It doesn't look so elegant, I would prefer to return them directly, and handle the error out of `schedulingCycle`. Diff likes https://github.com/kubernetes/kubernetes/commit/6d05fd0cb749c26709a977987358dd7e307352ac.

And maybe we can make `reason` part of ScheduleResult, it also makes sense to me if we really want to reduce the number of return values.

Reply:
I think both approaches have their complications.

Maybe this PR can just focus on splitting scheduleOne into schedulingCycle and bindingCycle in this PR (leave the calls to sched.FailureHandler where they were). Then, in a follow up issue people can present their approaches to refactoring and we pick the one that seems more reasonable.

Again, sorry for the back-and-forth. But sometimes it's hard to see the flaws of the approaches until they are fully fleshed out.

Reply:
👍 
Change it back in 7064e2b